firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/crcelements.txt

```diff
@@ -0,0 +1,131 @@
+      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
+      THEN("The body's payload section correctly stores the payload as '0x01 0x05 0x01 0x02 0x03 0x04 0x05'")
+      THEN("The output buffer is as expected")
+      # parse
+      THEN("The parse method reports ok status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
+      THEN("the payload given in the CRCElement constructor is independent of the input body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")
+      # write  
+      THEN("The second write method call returns ok status")
+      THEN("After the second write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
+      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
+      THEN("The body's payload section correctly stores the payload as '0x01 0x05 0x01 0x02 0x03 0x04 0x05'")
+      THEN("The output buffer is as expected")
+
+/* CRCElementReceiver
+Scenario: CRCElementReceiver correctly parses CRCElement bodies and performs consistency checking on them
+
+  GIVEN("A CRC element receiver of capacity 254 bytes")
+    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is given to the receiver")
+      THEN("the transform method reports ok status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
+      THEN("the payload buffer updated in the CRCElement is independent of the input body")
+
+    WHEN("A body without a complete 4-byte header and empty payload is given to the crc element receiver")
+      THEN("the transform status is equal to invalid parse")
+      THEN("the value returned by the crc accessor method is equal to 0x00000000r")
+      THEN("The body's payload section is empty")
+
+    WHEN("body with less than 4 bytes is given to the crc element receiver")
+      // body of length 1,2,3 bytes is passed, passing 4 bytes will give a invalid_crc error instead
+      THEN("the transform status is equal to invalid parse") // for each time
+      THEN("the value returned by the crc accessor method is equal to 0x00000000") // for each time
+
+    WHEN("A body with a complete 4-byte header, and a payload inconsistent with the CRC field of the header, is given to the receiver")
+      THEN("the transform method reports invalid crc status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
```

Comment:
This test should also have a THEN checking the value returned by the payload accessor method of output_crcelement.

Reply:
https://github.com/pez-globo/pufferfish-software/pull/281#discussion_r584449591

So when the GIVEN is a empty payload in output_crcelement, I'll say "returns empty payload buffer"
and for a GIVEN with payload in output_crcelement "returns 'payload'"

Reply:
I think this will be clearest if we split it into two THENs, as in https://github.com/pez-globo/pufferfish-software/pull/281#discussion_r585177976 - so here I think we'd assert that the contents of the payload buffer which had been given to the constructor are set to the bytes from the payload section of the input body, and then we'd assert that the payload accessor method returns a reference to that payload buffer in the constructor

Reply:
https://github.com/pez-globo/pufferfish-software/pull/281/files#diff-10ffb7b9ec39ab80c47393699e09e302b3ba3f871bf74fb4f834bf350cc97ef0R65

Added in CRCElement following https://github.com/pez-globo/pufferfish-software/pull/281#discussion_r585243812
in Receiver https://github.com/pez-globo/pufferfish-software/pull/281/files#diff-10ffb7b9ec39ab80c47393699e09e302b3ba3f871bf74fb4f834bf350cc97ef0R132